### DISCUSSION TITLE
PV Energy Total sensor values

### DISCUSSION THREAD

**Discussion by ac931274** [nathanmarlor/foxess_modbus#327] [General]
Earlier today I had to restart my FOXESS Hybrid/AC inverter. When I then started to get the values back from it the graphs were showing a massive -ve value when the readings had started coming through. 
<img width="940" alt="image" src="https://github.com/nathanmarlor/foxess_modbus/assets/318451/44213151-f4fc-4bc7-935b-ea7385f106a8">

I think that when the SensorDescription are being generated in  https://github.com/nathanmarlor/foxess_modbus/tree/main/custom_components/foxess_modbus/entities/entity_descriptions.py 
I think they should have the state_class of SensorStateClass.TOTAL_INCREASING and not state_class=SensorStateClass.TOTAL
<img width="441" alt="image" src="https://github.com/nathanmarlor/foxess_modbus/assets/318451/651dd685-b847-4848-a55f-4f2c86fe0864">
Above it shows that the incoming values were reset to 0 after the inverter restarted.

I think that it is the following function needs changing.  I am a C# developer and not a python one so this is my best guess.

def _pv_energy_total(key: str, models: list[EntitySpec], name: str, source_entity: str) -> EntityFactory:
    return ModbusIntegrationSensorDescription(
        key=key,
        models=models,
        device_class=SensorDeviceClass.ENERGY,
        state_class=SensorStateClass.TOTAL,
        native_unit_of_measurement="kWh",
        integration_method="left",
        name=name,
        source_entity=source_entity,
        unit_time=UnitOfTime.HOURS,
        icon="mdi:solar-power-variant-outline",
    )

If this is correct can it be fixed please?

**Comment by FozzieUK** [comment]
What version of HA are you running ?, the sensor should be ok as either Total or Total_increasing, but after a recent update to HA 2023.5.2 (3) the Riemann sums were reset (incorrectly) resulting from the sensor becoming an unknown or unavailable state - the net visual result is that you get a huge negative spike in the dashboard as the ‘sum’ resets to 0. It appears to have been fixed in 2023.5.4 and later which should stop it happening again.
I’ll leave it to @canton7 to comment on whether the state can/should be changed.

**Reply by ac931274** [reply]
I was running 2023.5.3 when the blib happened.  I have since updated to 2023.6.1 (HA Core)

**Reply by ac931274** [reply]
Is there anything I can do to fix the spike, or just learn to ignore the little things... :)

**Reply by canton7** [reply]
You can fix the current value easily enough by editing `.storage/core.restore_state` (with HA stopped). To fix past values you need to use SQL to edit your sqlite database, look at the `statistics` and `statistics_short_term` tables.

**Comment by canton7** [comment]
>I think they should have the state_class of SensorStateClass.TOTAL_INCREASING and not state_class=SensorStateClass.TOTAL

Why do you think that?

The integration sensors we use can go down as well as up (it's unlikely that they go down, but it's possible to get small negative values for things like PV1 Power). `total` should handle this properly and subtract from the sum, whereas `total_increasing` think that a meter cycle has restarted, and will *add* on the slightly lower value to the sum, effectively doubling the sum. [This page](https://developers.home-assistant.io/blog/2021/09/20/state_class_total/) has some good examples.

The default for Riemann integral sums is `total` (and in fact the line you found does nothing to change that... I'll fix).

Reading [the code](https://github.com/home-assistant/core/blob/b45659eb8441d43da06ee1c0bd350d139b98ced3/homeassistant/components/integration/sensor.py), Riemann sensors should never reset to zero, unless they are restored to a value which they can't parse [here](https://github.com/home-assistant/core/blob/b45659eb8441d43da06ee1c0bd350d139b98ced3/homeassistant/components/integration/sensor.py#L250) -- in that case you should get a warning message `<entity_id> could not restore last state ...`. Can you see whether you have this message in your log?

**Reply by ac931274** [reply]
Thanks for your quick reply. The reason that I thought it should be TOTAL_INCREASING is that that value coming from the INVERTER was always going up.  It was the resetting to 0 that caused the negative spike.

https://developers.home-assistant.io/docs/core/entity/sensor/ says the TOTAL the value can go up and down e.g. the level of a battery where as the those total PV values were since the inverter last reset.  This is shown in the values being recorded over time here:
<img width="1714" alt="image" src="https://github.com/nathanmarlor/foxess_modbus/assets/318451/28225f9b-46e6-4fb4-a1bf-7662ac1b5af1">
Right at the end you can see the value resetting back to 0 with the restart

I could well have got it wrong as I am only just getting started in to my HA involvement.  Many thanks for the work you are doing.

**Reply by canton7** [reply]
The value coming from the INVERTER is PV1 Power, which goes up and down depending on what the sun is doing. We then feed that into a [Riemann integral sensor](https://www.home-assistant.io/integrations/integration/) which sums the power over time to get total energy. (The inverter gives us totals for most things, but not PV1/PV2). The output of the Riemann integral *should* only go up, but because PV1 Power can occasionally go (slightly) negative, it can go (slightly) down as well.

The state_class affects the statistics maintained by HA. Storing every single reading is expensive, so after 10 days HA just maintains sets of buckets. There are two sets: 5 mins duration, and 1 hour duration. Each bucket holds the min, max, mean, and sum statistics (or a subset of these as appropriate), calculated from the raw values over the interval represented by the bucket. The state_class affects how the statistics are calculated, and in particular affects how the sum statistic is worked out. The point of `total_increasing` is that a decrease is interpreted as the meter resetting. If you hit the end of the month and your water meter resets from 1000 back to 0, you don't want to add -1000 to your sum statistic (which is what happens with `total`). However, with `total_increasing` a change from 1000 to 990 would be seen as an increase of 990, because we reset back to 0 and then added on 990; with `total`, the sum is just decreased by 10. If we changed the Riemann integral sensors to `total_increasing`, any small decrease in PV1 Energy Total caused by PV1 Power going negative would cause HA to (almost) double the sum statistic.

However, you're not even looking at the statistics here. (And even if you were, you're not going to care about the sum of total energy over time: you'll just look at the mean or max over each bucket). You're just looking at the latest value output of the Riemann integral sensor. 

Your problem here is that the Riemann integral sensor reset, out of the blue, which it should never do. This is most likely due to this issue https://github.com/home-assistant/core/pull/93674, which is now fixed.

(For what it's worth, I've been bitten by this as well and it sucks.)

**Reply by ac931274** [reply]
Thanks for your help.
